NF: Contour from ROI and Contour from Volume Introduced.

maharshi-gor opened the pull request:
### NF: Contour from ROI and Contour from Volume Introduced.

**PR Contents**
 Added function in utils to create a mesh from volume.
- Supports multiple objects in the volume.
- Requires an implementation bio-medical specific actor.
- Test cases for all the actors and utils function.
- Tutorial to try various combinations from volume.
- Fixed small bugs from surface actor.

skoudoro commented on fury/actor/bio.py:
```diff
@@ -129,3 +130,38 @@ def peaks_slicer(
+        The 3D volume data of shape (X, Y, Z).
+    affine : ndarray, optional
+        The affine transformation matrix to apply to the contour.
+    color : tuple, optional
```
the output color ? not clear

skoudoro commented on fury/actor/bio.py:
```diff
@@ -129,3 +130,38 @@ def peaks_slicer(
     return obj
+
+
+def contour_from_roi(
```
function name can be confusing since the input data does not seems to be a roi.

maharshi-gor replied:
I kept the old name. Should I update the variable name or the function name @skoudoro ?

skoudoro replied:
when you look at the old name and old function, look at this line:

https://github.com/fury-gl/fury/blob/792f89c8f09b840839a1a9137897b16b3d7678bd/fury/actor.py#L444

and look at this line : https://github.com/fury-gl/fury/blob/792f89c8f09b840839a1a9137897b16b3d7678bd/fury/actor.py#L468-L470

your function have the same name but seems to do something different and expect a different input. 
I need to introspect. Please, look at the old function also

maharshi-gor replied:
Hey @skoudoro 

I looked at the old function before, it was assuming that there is only one region of interest in the ROI. But the new object finds different regions not connected with each other.

The line #469 is scaling the labels to color range so the vtk can process them for resliceing.

The object relies mostly on vtk to do things for on the contrary we are doing things in pythonic way.

I can change the docstrings as they might be confusing.

Let me know if this makes sense.

skoudoro reviewed (changes requested):
Hi @maharshi-gor,

I need to play a bit with this but here the initial thoughts:

- the function inside `voxel_mesh_by_object` should be out, because they can be util.
- please, use more numpy, some operations can be avoided instead of a for loop

maharshi-gor committed:
NF: Contour from a volume introduced.

- Added APIs in utils to extract volume surface.
- Generation of Quads based on the axis and sign of coordinates.
- Contour Actor uses the surface actor internally.
- Small bugfix for the color_mode in the surface API.

maharshi-gor committed:
NF: Added Contour from ROI and Contour from label.

- These API are stacked on top of each other to reuse the apis.
- Provides flexibility to the user throw multiple rois in one mask.

maharshi-gor committed:
NF: Test cases for contour based actors.

- Introduced test cases for the utils apis.

maharshi-gor committed:
NF: Introduced tutorial for using contour based actors.

- Renamed the module from `surface` to `topology`.
- Test cases for contour_ actors.

skoudoro commented on fury/utils.py:
```diff
@@ -740,3 +772,207 @@ def get_transformed_cube_bounds(affine_matrix, vertex1, vertex2):
     return [min_vals, max_vals]
+
+
+def extract_surface_voxels(volume, label_value, *, structuring_element):
```
`=None`

skoudoro commented on fury/utils.py:
```diff
@@ -740,3 +772,207 @@ def get_transformed_cube_bounds(affine_matrix, vertex1, vertex2):
+        or a cropped sub-section.
+    label_value : int
+        The label identifying the object whose surface voxels should be returned.
+    structuring_element : ndarray
```
`, optional`

skoudoro commented on fury/utils.py:
```diff
@@ -740,3 +772,207 @@ def get_transformed_cube_bounds(affine_matrix, vertex1, vertex2):
+    """
+    object_mask = volume == label_value
+
+    surface_mask = object_mask & ~binary_erosion(
```
- `not`
- `&` vs`&&`
- clarify please

skoudoro commented on fury/utils.py:
```diff
@@ -740,3 +772,207 @@ def get_transformed_cube_bounds(affine_matrix, vertex1, vertex2):
+    labels, n = label(volume > 0, structure=struct)
+
+    if n == 0:
+        logging.info("No objects found in the volume.")
```
Todo later, uniform logger. sometimes print sometimes log

skoudoro commented on fury/utils.py:
```diff
@@ -740,3 +772,207 @@ def get_transformed_cube_bounds(affine_matrix, vertex1, vertex2):
+        occupied = np.zeros_like(valid, dtype=bool)
+        valid_idx = np.where(valid)
+        occupied[valid_idx] = object_mask[nz[valid_idx], ny[valid_idx], nx[valid_idx]]
+        exposed = ~occupied
```
same

skoudoro commented:
Can you fix the conflict here ? Thanks !

skoudoro reviewed (approved):
Thank you for this @maharshi-gor,

merging
